hooking up observers and clicking for ui node

TotalKrill opened the pull request:
Makes the newly merged picking usable for UI elements. 

currently it both triggers the events, as well as sends them as throught commands.trigger_targets. We should probably figure out if this is needed for them all. 

# Objective

Hooks up obserers and picking for a very simple example

## Solution

upstreamed the UI picking backend from bevy_mod_picking

## Testing

tested with the new example picking/simple_picking.rs


---

DasLixou commented on crates/bevy_ui/Cargo.toml:
```diff
@@ -26,6 +26,7 @@ bevy_reflect = { path = "../bevy_reflect", version = "0.15.0-dev", features = [
 bevy_render = { path = "../bevy_render", version = "0.15.0-dev" }
 bevy_sprite = { path = "../bevy_sprite", version = "0.15.0-dev" }
 bevy_text = { path = "../bevy_text", version = "0.15.0-dev", optional = true }
+bevy_picking= { path = "../bevy_picking", version = "0.15.0-dev", optional = true }
```
```suggestion
bevy_picking = { path = "../bevy_picking", version = "0.15.0-dev", optional = true }
```

DasLixou commented:
If there's an obvious use case for using events, I would keep them in, otherwise I would remove them for now. It's easier to add them later than removing a feature.

TotalKrill commented:
Did not know observesrs did not use events internally, but for now, removed the events that was not needed for further processing and changed them to Triggers instead. 

I agree that we can add them if needed.

jnhyatt commented on crates/bevy_internal/src/default_plugins.rs:
```diff
@@ -56,6 +56,12 @@ plugin_group! {
         bevy_gizmos:::GizmoPlugin,
         #[cfg(feature = "bevy_state")]
         bevy_state::app:::StatesPlugin,
+        #[cfg(feature = "bevy_picking")]
```
Plugin tuples can nest arbitrarily right? Just a nitpick (feel free to ignore), would be more concise if you throw these three into a tuple (or even a plugin group) and put the `#[cfg]` just once.

TotalKrill replied:
I fixed it into one single plugin, please resolve if you think this is enought.

jnhyatt commented on crates/bevy_picking/src/events.rs:
```diff
@@ -377,12 +383,9 @@ pub fn send_click_and_drag_events(
-    // Output
+    // Outputs used for further processing
     mut drag_map: ResMut<DragMap>,
-    mut pointer_click: EventWriter<Pointer<Click>>,
```
Why remove the `EventWriter`s here but not in `pointer_events`? Above it looks like you're both sending an event and triggering observers, but here only triggering observers

TotalKrill replied:
That is done because the events on other places are further processed and then turned into triggers here.

I removed the eventwriters that was not used for further processing within the picking crate

kristoff3r commented on crates/bevy_picking/src/lib.rs:
```diff
@@ -167,8 +176,27 @@ pub enum PickSet {
     Last,
 }
 
+/// One plugin that contains the [`input::InputPlugin`], [`PickingPlugin`] and the [`InteractionPlugin`],
+/// this is probable the plugin that will be most used.
+/// Note: for any of these plugins to work, they require a picking backend to be active,
+/// The picking backend is responsible to turn an input, into a [`crate::backend::PointerHits`]
+/// that [`PickingPlugin`] and [`InteractionPlugin`] will refine into Triggers
```
```suggestion
/// One plugin that contains the [`input::InputPlugin`], [`PickingPlugin`] and the [`InteractionPlugin`],
/// this is probably the plugin that will be most used.
/// Note: for any of these plugins to work, they require a picking backend to be active,
/// The picking backend is responsible to turn an input, into a [`crate::backend::PointerHits`]
/// that [`PickingPlugin`] and [`InteractionPlugin`] will refine into [`bevy_ecs::observer::Trigger`]s.
```

kristoff3r reviewed (approved):
LGTM, also tested the example on wasm

NthTensor reviewed (approved):
The changes to `bevy_picking/events` are acceptable. I imagine there will be a good amount of API churn there, and the event dispatchers will have to change significantly.

Good to have examples and basic back-ends working.

alice-i-cecile commented on crates/bevy_ui/src/picking_backend.rs:
```diff
@@ -0,0 +1,214 @@
+//! A picking backend for Ui nodes.
```
```suggestion
//! A picking backend for UI nodes.
```

alice-i-cecile commented on crates/bevy_ui/src/picking_backend.rs:
```diff
@@ -0,0 +1,214 @@
+//!
+//! ## Important Note
+//!
+//! This backend completely ignores [`FocusPolicy`](crate::FocusPolicy). The design of bevy ui's
```
```suggestion
//! This backend completely ignores [`FocusPolicy`](crate::FocusPolicy). The design of `bevy_ui`'s
```

alice-i-cecile commented on crates/bevy_ui/src/picking_backend.rs:
```diff
@@ -0,0 +1,214 @@
+//!
+//! ## Implementation Notes
+//!
+//! - Bevy ui can only render to the primary window
+//! - Bevy ui can render on any camera with a flag, it is special, and is not tied to a particular
+//!   camera.
+//! - To correctly sort picks, the order of bevy UI is set to be the camera order plus 0.5.
```
```suggestion
//! - `bevy_ui` can only render to the primary window
//! - `bevy_ui` can render on any camera with a flag, it is special, and is not tied to a particular
//!   camera.
//! - To correctly sort picks, the order of `bevy_ui` is set to be the camera order plus 0.5.
```

alice-i-cecile commented on crates/bevy_ui/src/picking_backend.rs:
```diff
@@ -0,0 +1,214 @@
+
+use bevy_picking::backend::prelude::*;
+
+/// Adds picking support for Ui nodes.
```
```suggestion
/// A plugin that adds picking support for UI nodes.
```

alice-i-cecile commented:
I'd like to improve the docs further (and unify this with the rest of `bevy_ui`), but this is a good step forward. Happy enough to merge this now.
